tasks/indexing/task_indexing.go

```diff
@@ -0,0 +1,298 @@
+
+	"github.com/ipfs/go-cid"
+	logging "github.com/ipfs/go-log/v2"
+	carv2 "github.com/ipld/go-car/v2"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-state-types/abi"
+
+	"github.com/filecoin-project/curio/harmony/harmonydb"
+	"github.com/filecoin-project/curio/harmony/harmonytask"
+	"github.com/filecoin-project/curio/harmony/resources"
+	"github.com/filecoin-project/curio/lib/indexing/indexstore"
+	"github.com/filecoin-project/curio/lib/passcall"
+	"github.com/filecoin-project/curio/lib/pieceprovider"
+
+	"github.com/filecoin-project/lotus/storage/sealer/storiface"
+)
+
+var log = logging.Logger("indexing")
+
+type IndexingTask struct {
+	db                   *harmonydb.DB
+	indexStore           *indexstore.IndexStore
+	pieceProvider        *pieceprovider.PieceProvider
+	maxCurrentForCluster int //TODO: Make this config
+}
+
+func NewIndexingTask(db *harmonydb.DB, indexStore *indexstore.IndexStore, pieceProvider *pieceprovider.PieceProvider) *IndexingTask {
+
+	return &IndexingTask{
+		db:            db,
+		indexStore:    indexStore,
+		pieceProvider: pieceProvider,
+	}
+}
+
+type itask struct {
+	UUID      string                  `db:"id"`
+	SpID      int64                   `db:"sp_id"`
+	Sector    abi.SectorNumber        `db:"sector_number"`
+	Proof     abi.RegisteredSealProof `db:"reg_seal_proof"`
+	PieceCid  string                  `db:"piece_cid"`
+	Size      abi.PaddedPieceSize     `db:"piece_size"`
+	CreatedAt time.Time               `db:"created_at"`
+	Offset    int64                   `db:"piece_offset"`
+	ChainID   abi.DealID              `db:"chain_deal_id"`
+	RawSize   int64                   `db:"raw_size"`
+}
+
+func (i *IndexingTask) Do(taskID harmonytask.TaskID, stillOwned func() bool) (done bool, err error) {
+
+	var tasks []itask
+
+	ctx := context.Background()
+
+	err = i.db.Select(ctx, &tasks, `SELECT 
+											mit.id as id, 
+											mit.sp_id as sp_id, 
+											mit.sector_number as sector_number, 
+											mit.piece_cid as piece_cid, 
+											mit.piece_size as piece_size, 
+											mit.created_at as created_at, 
+											mit.piece_offset as piece_offset, 
+											mit.reg_seal_proof as reg_seal_proof,
+											md.chain_deal_id as chain_deal_id,
+											mk.file_size as raw_size
+										FROM 
+											market_indexing_tasks mit
+										JOIN 
+											market_mk12_deal_pipeline mk ON mit.id = mk.uuid
+										JOIN 
+											market_mk12_deals md ON mk.id = md.uuid
+										WHERE 
+											mit.task_id = $1 
+										ORDER BY 
+											mit.created_at ASC;`, taskID)
+	if err != nil {
+		return false, xerrors.Errorf("getting indexing params: %w", err)
+	}
+
+	if len(tasks) != 1 {
+		return false, xerrors.Errorf("expected 1 sector params, got %d", len(tasks))
+	}
+
+	task := tasks[0]
+
+	// Check if piece is already indexed
+	var indexed bool
+	err = i.db.Select(ctx, &indexed, `SELECT indexed FROM market_piece_metadata WHERE piece_cid = $1`, task.PieceCid)
```

Comment:
Seems like we would start with either:
1. a scan of our local storage or
 2. triggers for incoming and shared storage (tasks that only one with that storage attached canAccept). 
then only process locally so that we never hit the network with anything but index data.